File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # Initialize a set to store characters in the current window
        char_set = set()
        # Initialize left pointer and max_length to track the longest substring
        left = 0
        max_length = 0

        # Iterate with the right pointer over the string
        for right in range(len(s)):
            # If the character is in the set, shrink the window by moving left
            while s[right] in char_set:
                char_set.remove(s[left])
                left += 1
            # Add the current character to the set
            char_set.add(s[right])
            # Update the max length of the substring found
            max_length = max(max_length, right - left + 1)

        return max_length

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) { // Sliding Window + Hash Table (ASCII Array)

        // approach 2
        // int left = 0, maxLength = 0;
        // HashSet<Character> charSet = new HashSet<>();

        // for (int right = 0; right < s.length(); right++) {

        // while (charSet.contains(s.charAt(right))) {
        // charSet.remove(s.charAt(left));
        // left++;
        // }

        // charSet.add(s.charAt(right));
        // maxLength = Math.max(maxLength, right - left + 1);
        // }
        // return maxLength;

        // approach 2

        int[] charIndex = new int[128]; // ASCII characters
        int maxLength = 0, left = 0;

        for (int right = 0; right < s.length(); right++) {

            left = Math.max(left, charIndex[s.charAt(right)]);
            maxLength = Math.max(maxLength, right - left + 1);

            charIndex[s.charAt(right)] = right + 1;
        }

        return maxLength;
    }
}